### ISSUE TITLE
Task module opening intermittently in personal context

### ISSUE THREAD

**Issue by taylorcjohnson** [MicrosoftDocs/msteams-docs#5346]
We were given feedback from Microsoft in an audit of our bot that we should have an actionable button in our welcome message sent in personal context to new users and when the bot is installed. The button invokes a task module (populated via adaptive card) that is part of our core functionality. However, nothing happens when the button is clicked in a personal context. A similar button that opens the same adaptive card will open the task module in a channel context.

No error message is visible to the user, but no action happens, and no `task/fetch` message is sent to the bot after clicking on it.
![Screen Shot 2022-02-25 at 11 01 31 AM](https://user-images.githubusercontent.com/10552296/155765127-3ec2d942-b899-43ff-9c53-d91a8c4258bf.png)

This seems very strongly related to this closed issue: https://github.com/MicrosoftDocs/msteams-docs/issues/4527 (we see the same console error when attempting in the browser: `TaskModuleLaunchService: launchTaskModuleByInvoke failed: "could not find app definition for app..."`).

The solution from that closed issue indicates that every user would have to install the app before the task module would work. But if our welcome message is sent when our bot is added to a Team's channel, we cannot guarantee that the users would have already installed the bot. 

Is there a preferred way to have companies or customers install the bot so they can always use task modules in every context?

Is there a way to ensure that users who have not installed the bot separately, but have access to it in a channel, are still be able to use task modules in a personal context?

**tl;dr:** task modules won't always open in personal context, but with no errors or messages sent to the bot.

**Comment by Nivedipa-MSFT** [comment]
@taylorcjohnson - We are investigating your query. We will get back to you once we have any update.

**Comment by taylorcjohnson** [comment]
@Nivedipa-MSFT Hello, I just wanted to follow-up. We appreciate any recommendations.

**Comment by Nivedipa-MSFT** [comment]
@taylorcjohnson - You can use the **Microsoft Graph API to proactively install your bot for your users**. Refer below document for more details.
Ref Doc: https://docs.microsoft.com/en-us/microsoftteams/platform/graph-api/proactive-bots-and-messages/graph-proactive-bots-and-messages?tabs=dotnet

Using below graph API you can determine whether your bot is currently installed for a message recipient.
`GET https://graph.microsoft.com/v1.0/users/{user-id}/teamwork/installedApps?$expand=teamsApp&$filter=teamsApp/id eq '{teamsAppId}'`
Ref Doc: https://docs.microsoft.com/en-us/microsoftteams/platform/graph-api/proactive-bots-and-messages/graph-proactive-bots-and-messages?tabs=dotnet#determine-whether-your-bot-is-currently-installed-for-a-message-recipient

**Comment by taylorcjohnson** [comment]
Thank you @Nivedipa-MSFT! We are going to try to quickly implement this approach and see if it resolves our issues with task modules in personal contexts. I'll reply with what we find out. Thank you again for the recommendation.
